### PR TITLE
feat(component): restrict actions and header from modals

### PR DESCRIPTION
**Pull request by jorgemoya**
# BREAKING CHANGES

## What
Headers and Actions are now a prop of `modal` component. Visually no change.

Added `header` & `actions` props.

`actions` accepts an array of `Objects` with `Button` props and a `text` prop.

Example:
```ts
[
  { text: 'Cancel', onClick: () => setIsOpen(false) },
  { text: 'Apply', onClick: () => setIsOpen(false) },
]
```

Updated tests and docs.

### CODE REVIEW

File: packages/big-design/src/components/Modal/Modal.tsx [hunk 1]
```diff
@@ -16,11 +16,15 @@ import {
...
 export interface ModalProps {
+  actions?: React.ReactNode;
   backdrop: boolean;
-  isOpen: boolean;
   closeOnClickOutside: boolean;
   closeOnEscKey: boolean;
+  header?: string;
+  isOpen: boolean;
   variant: 'modal' | 'dialog';
+  withActionsBorder?: boolean;
```
Comment (deini) [inline packages/big-design/src/components/Modal/Modal.tsx]: Do we have multiple designs? I think this should be consistent in all modals and not configurable.

File: packages/big-design/src/components/Modal/Modal.tsx [hunk 2]
```diff
@@ -136,18 +116,28 @@ export class Modal extends React.PureComponent<ModalProps, ModalState> {
...
-    return React.Children.map(children, child => {
-      if (React.isValidElement(child) && child.type === Modal.Header) {
-        return React.cloneElement(child as React.ReactElement<HeadingProps>, {
-          id: this.headerUniqueId,
-        });
-      }
+    return (
+      header && (
+        <StyledModalHeader id={this.headerUniqueId} withBorder={withHeaderBorder}>
+          <H2 margin="none">{header}</H2>
```
Comment (deini) [inline packages/big-design/src/components/Modal/Modal.tsx]: Should we do a runtime check to make sure `header` is a string? 🤔

**Comment by deini** [comment]
Can you also add the `BREAKING CHANGE` section to your commit body with examples of how to migrate? 🙏

File: packages/big-design/src/components/Modal/Modal.tsx (hunk 1 above)
Comment (chanceaclark) [reply packages/big-design/src/components/Modal/Modal.tsx]: @deini It seems like you added it in #133 to support @animesh1987 use case. Is this something we need to touch back with design?

File: packages/big-design/src/components/Modal/Modal.tsx [hunk 3]
```diff
@@ -136,18 +116,28 @@ export class Modal extends React.PureComponent<ModalProps, ModalState> {
...
-      return child;
-    });
+    return (
+      actions && (
+        <StyledModalActions justifyContent="flex-end" withBorder={withActionsBorder}>
+          {actions}
+        </StyledModalActions>
+      )
+    );
   }
```
Comment (chanceaclark) [inline packages/big-design/src/components/Modal/Modal.tsx]: To go along with @deini runtime check for strings, should we be consistent and do the same for `Buttons` within this `renderActions` method?

File: packages/big-design/src/components/Modal/spec.tsx [hunk 4]
```diff
@@ -211,3 +213,40 @@ test('body has scroll locked on modal open', () => {
...
+test('renders header', () => {
+  const { getByText } = render(<Modal isOpen={true} header="Header Title" />);
+
+  expect(getByText('Header Title')).toBeInTheDocument();
+});
+
+test('renders header border', () => {
+  render(<Modal isOpen={true} header="Header Title" withHeaderBorder={true} />);
+
+  expect(document.body).toMatchSnapshot();
```
Comment (chanceaclark) [inline packages/big-design/src/components/Modal/spec.tsx]: 🍹You may also want to check if it has `toHaveStyleRule`.

File: packages/big-design/src/components/Modal/spec.tsx [hunk 5]
```diff
@@ -211,3 +213,40 @@ test('body has scroll locked on modal open', () => {
...
+  const actions = (
+    <>
+      <Button variant="subtle">Cancel</Button>
+      <Button>Apply</Button>
+    </>
+  );
+
+  render(<Modal isOpen={true} actions={actions} withActionsBorder={true} />);
+
+  expect(document.body).toMatchSnapshot();
```
Comment (chanceaclark) [inline packages/big-design/src/components/Modal/spec.tsx]: 🍹Same as above

File: packages/docs/PropTables/ModalPropTable.tsx [hunk 6]
```diff
@@ -4,11 +4,8 @@ import { Code, PropTable } from '../components';
 
 export const ModalPropTable: React.FC = () => (
   <PropTable>
-    <PropTable.Prop name="isOpen" types="boolean" required>
-      Determines if the modal/dialog is open.
-    </PropTable.Prop>
-    <PropTable.Prop name="onClose" types="() => void" required>
-      Function that will be called on close events.
+    <PropTable.Prop name="actions" types="React.ReactNode">
+      Sets the actions of the modal.
```
Comment (chanceaclark) [inline packages/docs/PropTables/ModalPropTable.tsx]: Maybe we should reword this to what _should_ be passed into the actions (i.e. `Button`'s).

File: packages/docs/PropTables/ModalPropTable.tsx [hunk 7]
```diff
@@ -19,24 +16,23 @@ export const ModalPropTable: React.FC = () => (
     <PropTable.Prop name="closeOnEscKey" types="boolean" defaults="true">
       Controls whether <Code>onClose</Code> is called when pressing the ESC key.
     </PropTable.Prop>
+    <PropTable.Prop name="header" types="string">
+      Header title of the modal.
```
Comment (chanceaclark) [inline packages/docs/PropTables/ModalPropTable.tsx]: 🤷‍♀️Maybe different verbiage, but along these lines?
```suggestion
      Sets visible text that describes the content of the modal.
```

**Comment by jorgemoya** [comment]
> Can you also add the `BREAKING CHANGE` section to your commit body with examples of how to migrate? 🙏

Is this done when you squash and merge the commit?

**Comment by chanceaclark** [comment]
> > Can you also add the `BREAKING CHANGE` section to your commit body with examples of how to migrate? 🙏
> 
> Is this done when you squash and merge the commit?

Yeah, you can write it in full markdown.

File: packages/big-design/src/components/Modal/Modal.tsx (hunk 3 above)
Comment (jorgemoya) [reply packages/big-design/src/components/Modal/Modal.tsx]: Makes sense however how can you test for `Buttons` when you receive a set of actions?

For example in:
```
  const actions = (
    <>
      <Button variant="subtle">Cancel</Button>
      <Button>Apply</Button>
    </>
  );
```
In this case `React.Children.map` wouldn't iterate over the buttons.

File: packages/big-design/src/components/Modal/Modal.tsx (hunk 1 above)
Comment (jorgemoya) [reply packages/big-design/src/components/Modal/Modal.tsx]: Removed borders.

File: packages/docs/PropTables/ModalPropTable.tsx [hunk 8]
```diff
@@ -1,14 +1,11 @@
...
 export const ModalPropTable: React.FC = () => (
   <PropTable>
-    <PropTable.Prop name="isOpen" types="boolean" required>
-      Determines if the modal/dialog is open.
-    </PropTable.Prop>
-    <PropTable.Prop name="onClose" types="() => void" required>
-      Function that will be called on close events.
+    <PropTable.Prop name="actions" types="object[]">
+      Accepts an array of objects with <NextLink href="/button">Button</NextLink> props and a <Code>text</Code> label.
     </PropTable.Prop>
```
Comment (jorgemoya) [inline packages/docs/PropTables/ModalPropTable.tsx]: What do you guys think about this prop description?

File: packages/big-design/src/components/Modal/Modal.tsx [hunk 9]
```diff
@@ -29,42 +31,20 @@ interface ModalState {
...
 }
 
-export interface ModalActionsProps {
-  withBorder?: boolean;
-}
-
-export interface ModalHeaderProps {
-  withBorder?: boolean;
+interface Action extends Omit<ButtonProps, 'children'> {
+  text?: string;
```
Comment (deini) [inline packages/big-design/src/components/Modal/Modal.tsx]: Should this be required?

File: packages/big-design/src/components/Modal/Modal.tsx [hunk 10]
```diff
@@ -29,42 +31,20 @@ interface ModalState {
   modalContainer: HTMLDivElement | null;
 }
 
-export interface ModalActionsProps {
-  withBorder?: boolean;
-}
-
-export interface ModalHeaderProps {
-  withBorder?: boolean;
+interface Action extends Omit<ButtonProps, 'children'> {
```
Comment (deini) [inline packages/big-design/src/components/Modal/Modal.tsx]: What do you think about renaming this `ModalAction`? We also need to export it.

File: packages/big-design/src/components/Modal/Modal.tsx [hunk 11]
```diff
@@ -136,18 +118,34 @@ export class Modal extends React.PureComponent<ModalProps, ModalState> {
...
-        });
-      }
+  private renderActions() {
+    const { actions } = this.props;
 
-      return child;
-    });
+    return (
+      actions &&
+      Array.isArray(actions) && (
```
Comment (deini) [inline packages/big-design/src/components/Modal/Modal.tsx]: I don't think we need the initial `actions &&`

File: packages/big-design/src/components/Modal/Modal.tsx (hunk 9 above)
Comment (jorgemoya) [reply packages/big-design/src/components/Modal/Modal.tsx]: I was thinking we could have only icon buttons

File: packages/big-design/src/components/Modal/Modal.tsx (hunk 10 above)
Comment (jorgemoya) [reply packages/big-design/src/components/Modal/Modal.tsx]: Why export it?

File: packages/big-design/src/components/Modal/Modal.tsx (hunk 9 above)
Comment (deini) [reply packages/big-design/src/components/Modal/Modal.tsx]: Good point, can't think of it being a good use case but let's leave it for now. Good catch.

File: packages/big-design/src/components/Modal/Modal.tsx (hunk 10 above)
Comment (deini) [reply packages/big-design/src/components/Modal/Modal.tsx]: For example, if you don't want to inline the `actions` prop you could do.

```tsx
const actions: ModalAction[] = [
  { text: 'Cancel', onClick: () => {} }
];

// And use it like
<Modal actions={actions} />
```

File: packages/docs/pages/Modal/ModalPage.tsx [hunk 12]
```diff
@@ -26,29 +26,28 @@ export default () => (
...
-                </Text>
-              </Modal.Body>
-
-              <Modal.Actions>
-                <Button variant="subtle" onClick={() => setIsOpen(false)}>
-                  Cancel
-                </Button>
-                <Button onClick={() => setIsOpen(false)}>Apply</Button>
-              </Modal.Actions>
+            <Modal
```
Comment (icatalina) [inline packages/docs/pages/Modal/ModalPage.tsx]: This looks great! 👍
